File: app/src/main/java/com/jakebarnby/filemanager/sources/onedrive/OneDriveLoaderTask.java
package com.jakebarnby.filemanager.sources.onedrive;

import com.jakebarnby.filemanager.sources.LoaderTask;
import com.jakebarnby.filemanager.sources.SourceListener;
import com.jakebarnby.filemanager.sources.models.Source;
import com.jakebarnby.filemanager.sources.models.SourceFile;
import com.jakebarnby.filemanager.util.TreeNode;
import com.microsoft.graph.extensions.DriveItem;
import com.microsoft.graph.extensions.IDriveItemCollectionPage;
import com.microsoft.graph.http.GraphServiceException;

import java.util.List;

/**
 * Created by Jake on 8/2/2017.
 */

public class OneDriveLoaderTask extends LoaderTask {

    private DriveItem mRootDriveItem;

    public OneDriveLoaderTask(Source source, SourceListener listener, DriveItem rootDriveItem) {
        super(source, listener);
        this.mRootDriveItem = rootDriveItem;
    }

    @Override
    protected Object initRootNode(String path) {
        SourceFile rootSourceFile = new OneDriveFile(mRootDriveItem);
        rootSourceFile.setDirectory(true);
        mRootTreeNode = new TreeNode<>(rootSourceFile);
        mCurrentNode = mRootTreeNode;
        mSource.setCurrentDirectory(mRootTreeNode);
        mSource.setQuotaInfo(OneDriveFactory.getInstance().getStorageStats());
        return mRootDriveItem;
    }

    @Override
    protected TreeNode<SourceFile> readFileTree(Object rootObject) {
        try {
            if (rootObject != null && rootObject instanceof DriveItem) {
                DriveItem node = (DriveItem) rootObject;
                IDriveItemCollectionPage items = OneDriveFactory
                        .getInstance()
                        .getGraphClient()
                        .getMe()
                        .getDrive()
                        .getItems(node.id)
                        .getChildren()
                        .buildRequest()
                        .select("id,name,webUrl,folder,size,createdDateTime,lastModifiedDateTime")
                        .expand("thumbnails")
                        .get();

                List<DriveItem> pageItems = items.getCurrentPage();
                long dirSize = 0L;
                if (pageItems != null) {
                    for (DriveItem file : pageItems) {
                        SourceFile sourceFile = new OneDriveFile(file);
                        if (file.folder != null) {
                            mCurrentNode.addChild(sourceFile);
                            mCurrentNode = mCurrentNode.getChildren().get(mCurrentNode.getChildren().size() - 1);
                            readFileTree(file);
                            mCurrentNode.getParent().getData().addSize(mCurrentNode.getData().getSize());
                            mCurrentNode = mCurrentNode.getParent();
                        } else {
                            dirSize += file.size;
                            mCurrentNode.addChild(sourceFile);
                        }
                    }
                    mCurrentNode.getData().addSize(dirSize);
                }
            }
        } catch (GraphServiceException e) {
            mListener.onLoadError(e.getMessage() != null ? e.getMessage() : "");
            mSucess = false;
        }
        return mRootTreeNode;
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/sources/onedrive/OneDriveLoaderTask.kt
package com.jakebarnby.filemanager.sources.onedrive

import com.jakebarnby.filemanager.workers.LoaderTask
import com.jakebarnby.filemanager.models.Source
import com.jakebarnby.filemanager.models.SourceFile
import com.jakebarnby.filemanager.sources.dropbox.DropboxClient
import com.jakebarnby.filemanager.ui.sources.SourceFragmentContract
import com.jakebarnby.filemanager.util.TreeNode
import com.microsoft.graph.extensions.DriveItem
import com.microsoft.graph.http.GraphServiceException
import javax.inject.Inject

/**
 * Created by Jake on 8/2/2017.
 */
class OneDriveLoaderTask(
    source: Source,
    presenter: SourceFragmentContract.Presenter,
    private val rootDriveItem: DriveItem
) : LoaderTask(source, presenter) {

    @Inject
    lateinit var oneDriveClient: OneDriveClient

    override fun initRootNode(path: String): Any? {
        val rootSourceFile: SourceFile = OneDriveFile(rootDriveItem)
        rootSourceFile.isDirectory = true
        rootTreeNode = TreeNode(rootSourceFile)
        currentNode = rootTreeNode
        source.currentDirectory = rootTreeNode
        source.setQuotaInfo(oneDriveClient.storageInfo)
        return rootDriveItem
    }

    override fun readFileTree(rootObject: Any?): TreeNode<SourceFile> {
        try {
            if (rootObject == null || rootObject !is DriveItem) {
                return rootTreeNode
            }

            var dirSize = 0L
            var items = oneDriveClient.getFilesByParentId(rootObject.id)

            if (items?.currentPage == null) {
                return rootTreeNode
            }

            while (items != null) {
                val pageItems = items.currentPage

                for (file in pageItems) {
                    val sourceFile = OneDriveFile(file)
                    if (file.folder == null) {
                        dirSize += file.size
                        currentNode.addChild(sourceFile)
                        continue
                    }

                    currentNode.addChild(sourceFile)
                    currentNode = currentNode.children[currentNode.children.size - 1]
                    readFileTree(file)

                    currentNode.parent!!.data.size += currentNode.data.size
                    currentNode = currentNode.parent!!
                }
                items = oneDriveClient.getNextPage(items)
            }
            currentNode.data.size += dirSize
        } catch (e: GraphServiceException) {
            presenter.onLoadError(if (e.message != null) e.message else "")
            success = false
        }
        return rootTreeNode
    }

}